test: enable upgrade to image built from source

pawelpaszki opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Enable upgrade to image built from source. if `KUBERAY_OPERATOR_UPGRADE_IMAGE` is not provided, empty default will be returned and the e2e upgrade will progress without any changes. In case of providing `KUBERAY_OPERATOR_UPGRADE_IMAGE` with tag `nightly`, operator will be upgraded to the nightly image built from source.

Note that in order for the test to work, image needs to be built and loaded into a kind cluster (by GH actions or equivalent workflow). 

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
  - [x] e2e tests were executed with the original flow and with using built nightly image - both were successful
  - [x] Manual tests
  - [ ] This PR is not tested :(

kevin85421 commented:
cc @ryanaoleary @andrewsykim would you mind reviewing this change? Thanks.

ryanaoleary commented on ray-operator/test/e2eupgrade/rayservice_operator_upgrade_test.go:
```diff
@@ -30,9 +35,20 @@ func TestZeroDowntimeUpgradeAfterOperatorUpgrade(t *testing.T) {
 	namespace := test.NewTestNamespace()
 	rayServiceName := "rayservice-sample"
 
+	kuberayOperatorUpgradeImage := GetKubeRayOperatorUpgradeImage()
```
Why do we need to add and detect another environment variable here, can we just use the existing `KUBERAY_TEST_UPGRADE_IMAGE`?

pawelpaszki replied:
I didn't want to limit the upgrade to the same image (by specifying full image), hence this change, but I can change it to only provide the version

pawelpaszki replied:
sorry.  I was busy with other stuff. can you take another look please?

ryanaoleary commented on ray-operator/test/support/environment.go:
```diff
@@ -36,6 +39,12 @@ func GetRayImage() string {
 
+// GetKubeRayOperatorUpgradeImage - returns upgrade image from env var or empty default
+// (to prevent from unwanted upgrades in the e2e tests)
+func GetKubeRayOperatorUpgradeImage() string {
```
Same comment as above, can we just use the existing `GetKubeRayUpgradeVersion` and if it's specified to the nightly image, then build from source?

ryanaoleary reviewed (approved):
LGTM!

kevin85421 commented on ray-operator/test/e2eupgrade/rayservice_operator_upgrade_test.go:
```diff
@@ -76,15 +82,32 @@ func TestZeroDowntimeUpgradeAfterOperatorUpgrade(t *testing.T) {
-	g.Expect(err).NotTo(HaveOccurred())
-	g.Eventually(cmd, TestTimeoutShort).Should(WithTransform(ProcessStateSuccess, BeTrue()))
+	test.T().Logf("Upgrading the KubeRay operator to %s", upgradeVersion)
+	if upgradeVersion == nightlyUpgradeImageVersion {
```
Can you use the Helm chart on the same commit instead of `kubectl set image`? We can also add the upgrade test from the latest release to the nightly commit in the CI in a follow up PR.

pawelpaszki replied:
I don't think helm can be used, because the `nightly` isn't available there (if I understand your question correctly)

pawelpaszki replied:
@kevin85421 are you ok with leaving the code as is and merging this PR if you are happy?

pawelpaszki commented:
@kevin85421 can you take a look at my latest comment please?

kevin85421 commented:
Btw, sorry for the late review, I am OOO this week.

kevin85421 commented:
Are you on the Slack? You can send me a DM message and we can discuss how to streamline your contribution experience next time.

pawelpaszki commented:
@kevin85421 I think I have addressed your comments. can you take another look please? as for slack - I am on my company slack. can you send me an email with your workspace, I will try to join it and contact you there

kevin85421 commented:
@pawelpaszki https://github.com/ray-project/ray there is a slack link in the README.
